pyanaconda/modules/subscription/subscription.py

```diff
@@ -174,3 +240,80 @@ def set_system_purpose_data(self, system_purpose_data):
         self._system_purpose_data = system_purpose_data
         self.system_purpose_data_changed.emit()
         log.debug("System purpose data set to %s.", system_purpose_data)
+
+    # subscription request
+
+    @property
+    def subscription_request(self):
+        """Subscription request.
+
+        A DBus structure holding data to be used to subscribe the system.
+
+        :return: subscription request DBus structure
+        :rtype: DBusData instance
+        """
+        # start by doing a deep copy so that we don't change the "internal"
+        # structure when we clear private data
+        request_copy = copy.deepcopy(self._subscrition_request)
+
+        # clear private data
+        # - we don't need to set private data indicators as set_subscription_data()
+        #   handles that for us
+        request_copy.account_password = ""
+        request_copy.activation_keys = []
+        request_copy.server_proxy_password = ""
+
+        return request_copy
+
+    def set_subscription_request(self, subscription_request):
+        """Set a subscription request.
+
+        Set the complete DBus structure containing subscription
+        request data.
+
+        :param subscription_request: subscription request structure to be set
+        :type subscription_request: DBus structure
+        """
+        current_request = copy.deepcopy(self._subscrition_request)
+
+        # set private data indicators
+        subscription_request.account_password_set = bool(
+                subscription_request.account_password
+        )
+        subscription_request.activation_keys_set = bool(
+                subscription_request.activation_keys
+        )
+        subscription_request.server_proxy_password_set = bool(
+                subscription_request.server_proxy_password
+        )
```

Comment:
These properties will be used by the UI to check if a password/key has been set previously outside of the UI, via kickstart DBus or similar.

As for setting the value of the property in the DBus structure itself - I guess we could implement that & possibly drop the public setter for the _set properties hen we are at it.

But we still need to validate this when we set the structure to the module as well, otherwise in a case of a client not written in Python that does not have our SubscriptionRequest class available we can't really assure it will all be set correctly & should not blindly accept the data into the module.

Reply:
I think that the design with the boolean property `<secret>_set` is not going to work in DBus structures, because the value of the property can be interpreted in different ways and confuse everyone. It should be more obvious what is going on.

I suggest to use a property `<secret>_type` that can be set to values `NONE`, `HIDDEN` and `TEXT`. The service will return `NONE`, when the secret is not set, and `HIDDEN`, when the secret is set, but its value is not provided in the property `<secret>`. The client will return `NONE`, when the secret should be cleared, `HIDDEN`, when the hidden value of the secret should be used, and `TEXT`, when the plain text value of the property `<secret>` should be used. This design could be used also in the Users module. Types can be extended with `ENCRYPTED`.

Then the structure could have a method called `set_<secret>` with an argument `secret`. If the argument `secret` is None, then the type will be set to `NONE` and the secret to an empty string. Otherwise, we will set the type to `TEXT` and set the secret. This will be useful for the client and the service.

The service could use a method called `hide_<secret>`, that will change the type of the secret from `TEXT` to `HIDDEN` and set the secret to an empty string.

Ideally, we should have a DBus structure called `SecretData` that can be shared by all modules that provide secrets. We support nested DBus structures, so you would have just a property called for example `server_proxy_password` that would return an instance of `SecretData`. The problem is with the list of activation keys. I will have to think about it a little more.

Reply:
https://github.com/rhinstaller/anaconda/pull/2412

Reply:
Sounds perfect, let's use that! :) And thanks a lot! :)